### PR TITLE
TextPainter should dispatch creation and disposal events.

### PR DESCRIPTION
**Pull request by ksokolovskyi**
### Description
- Adds `TextPainter` creation and disposal dispatching for memory leak tracking as part of https://github.com/flutter/flutter/issues/137311.

### Tests
- Updates `text_painter_test.dart` to test `TextPainter` object creation and disposal dispatching;
- Fixes memory leak in `_SnackBarState`;
- Fixes memory leak in `_SwitchPainter`;
- Fixes memory leak in `_FlutterLogoPainter`;
- Fixes memory leak in `RenderDecoratedBox`;
- Fixes memory leak in `BannerPainter`.

## Pre-launch Checklist

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [X] I signed the [CLA].
- [X] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [X] I added new tests to check the change I am making, or this PR is [test-exempt].
- [X] All existing and new tests are passing.

### CODE REVIEW

**Comment by ksokolovskyi** [comment]
cc @polina-c

File: packages/flutter/lib/src/painting/flutter_logo.dart [hunk 1]
```diff
@@ -219,6 +219,12 @@ class _FlutterLogoPainter extends BoxPainter {
   late TextPainter _textPainter;
   late Rect _textBoundingRect;
 
+  @override
+  void dispose() {
+    _textPainter.dispose();
```
Comment (navaronbracke) [inline packages/flutter/lib/src/painting/flutter_logo.dart:224]: Now the TODO on line 218 is redundant?

File: packages/flutter/test/painting/text_painter_test.dart [hunk 2]
```diff
@@ -1525,6 +1526,13 @@ void main() {
         expect(metrics, hasLength(1));
     }
   }, skip: kIsWeb && !isCanvasKit); // [intended] Browsers seem to always round font/glyph metrics.
+
+  test('TextPainter dispatches memory events', () async {
+    await expectLater(
+      await memoryEvents(() => TextPainter().dispose(), TextPainter),
+      areCreateAndDispose,
```
Comment (navaronbracke) [inline packages/flutter/test/painting/text_painter_test.dart:1533]: @polina-c Probably an unfortunate typo in `leak_tracker_flutter_testing` but shouldn't this be named as `areCreatedAndDisposed` ?

**Review by navaronbracke — COMMENTED** [review]
Just skimming through (I tend to keep an eye on leak tracker things :) )

File: packages/flutter/lib/src/painting/text_painter.dart [hunk 3]
```diff
@@ -498,7 +500,15 @@ class TextPainter {
        _locale = locale,
        _strutStyle = strutStyle,
        _textWidthBasis = textWidthBasis,
-       _textHeightBehavior = textHeightBehavior;
+       _textHeightBehavior = textHeightBehavior {
+    if (kFlutterMemoryAllocationsEnabled) {
```
Comment (polina-c) [inline packages/flutter/lib/src/painting/text_painter.dart:504]: Can you add TODO here: https://github.com/flutter/flutter/issues/137435

File: packages/flutter/test/painting/text_painter_test.dart (hunk 2 above)
Comment (polina-c) [reply packages/flutter/test/painting/text_painter_test.dart:1533]: nope
the name is about events
the name states that the events are about creation and disposal, not that they are created and disposed

**Comment by ksokolovskyi** [comment]
@polina-c, @navaronbracke thanks for the review. 
I made some adjustments according to the latest comments. Could you please take a look again?

**Review by polina-c — APPROVED** [review]

File: packages/flutter/test/painting/text_painter_test.dart (hunk 2 above)
Comment (navaronbracke) [reply packages/flutter/test/painting/text_painter_test.dart:1533]: I see. That indeed makes more sense.

**Pull request merged**